### ISSUE TITLE
Unexpected callback argument type

### ISSUE THREAD

**Issue by BrandonZacharie** [barrysteyn/node-scrypt#63]
In version 4.0.6 `verifyHash` may invoke its `callback` argument with the first argument `err` as `undefined`. However, the documentation states that this argument _is defined as type `Error`_. In JavaScript, being defined means that it (excluding primitives) should be `null` in the absence of an instance as opposed to `undefined`. In essence, `undefined` is short for _undefined behavior_ which is generally not a good thing. I use strict equality comparisons (i.e. `err === null`) for performance (i.e. skip type coercion) and readability (i.e. imply a reference type); this behavior prevents that.

**Comment by barrysteyn** [comment]
@BrandonZacharie This is the problem with JavaScript in general. IMHO, a language where people quibble over the meaning of undefined vs null is a bit broken (don't worry, I still like JavaScript, I just think it needs to be fixed).

What you say makes sense, can I ask that if I change this to Null, will you update the documentation where necessary?

**Comment by BrandonZacharie** [comment]
I understand how you feel; something feels evil about having both especially since `undefined == null`. Unfortunately, `undefined` is a necessary evil. Since JavaScript is not a strongly typed language `var foo;` doesn't tell you if `foo` is supposed to represent a struct (which can never be `null`), a class instance, an integer, or anything else for that matter. In other words, we don't know what to do with it. The language gives us `undefined` as a means to represent something for which we don't know with what to do. it's good and bad but it's only good when it lets us know that we were given something useless. For example, as demonstrated in the first call below, it let's us know when we aren't given an argument anything at all.

```
function bar(a, b) { return a + b; }

bar(1);           // NaN ... runtime says: ummm... I don't know what that is but it's definitely not a number.
bar(undefined, 1) // NaN ... runtime says: defined as undefined is still undefined... nice try.
bar(0, null)      // 0 ... runtime says: hmmm... null is a type of something... let me hear what Number has to say about this.
bar(null, 1)      // 1 ... runtime says: Number said that null is equal to zero... so, 0 + 1 = 1.... done.
```
What if an argument is optional? How will we know when `b` is not provided? Sure, we can look at `arguments.length` but how will we know when `a` is not provided? We'd be in trouble without `undefined` as we would be unable to determine caller intent.

Long story short, `null` is a type and `undefined` is not. `undefined` is something we should only ever see enter a scope by accident with few exceptions.

You're probably more than well aware of all of this. I share this because it's these kinds of examples that prevent me from (╯°□°)╯︵ ┻━┻ when I come across something seemingly funky in JavaScript.

Finally, I'd like to say thank you on behalf of the many people I know, myself included, who appreciate every second you have put in to this project. I would be more than happy to contribute in any way including, but not limited to, documentation.

**Comment by barrysteyn** [comment]
Hi @BrandonZacharie 

Thanks for the recap regarding undefined and null, it brings me back to my "JavaScript the good parts" days :) What a classic book!

The thing is, even though I get you regarding why undefined is used, the fact that both undefined and null can - by and large - be used interchangeably and (again - by and large -) have the same effect means it is just badly designed. Other languages that are not strongly typed (think Python) do not have both an undefined and null "meanings". Okay, I hear you that JavaScript does some special things like being able to omit function arguments in any order, but honestly, couldn't the language makers have made null do that as well! So null represents an empty object, why can't it represent the absence of anything as well.

I want to end off this rant by saying that you made a very logical and coherent argument above, and I absolutely understand what undefined is doing. I just don't think Brandon Eich thought about it very long and hard before he made JavaScript. If you think about it, the examples you have given above in your argument, you could get away with only using null (or only using undefined if that is your poison).

Thanks so much for the offer regarding programming? Do you have any ideas regarding anything you would like to add to this module? I would be keen to entertain any ideas...

**Comment by BrandonZacharie** [comment]
You're right that, in practice, they're quite interchangeable. With the exception of a few legitimate cases, my argument was largely based on principle. I'll leave the full explanation of why we can't use `null` for everything to another day. FWIW, I have _never_ seen a JavaScript built-in or native Node.js module intentionally pass `undefined`.

> Do you have any ideas regarding anything you would like to add to this module?

I think you've made excellent design decisions regarding the scope of this module. The only ideas I have concern documentation. I'd be happy to have a discussion about them in a separate issue.

**Comment by barrysteyn** [comment]
HI Brandon

I have made a brandon branch which should return null instead of undefined for all error objects. I was hoping you could do the following:

 1. Test it!
 2. If it works, maybe you could write some tests and add it to the tests (although you don't have to if you don't want to).
 3. Add some documentation. Do note that not only verify will throw an error if you get the encodings incorrect, but so will everything else.
 4. Can you please link to the Node rules and guidelines in the docs so that people don't bring up this issue again for debate (and mostly, so if I forget, I can just look at it over :) )

Thanks so much helping out, I really appreciate it.

**Comment by BrandonZacharie** [comment]
I couldn't get around to it yesterday but I'll take some time to handle it this weekend.

**Comment by barrysteyn** [comment]
Okay, thanks for letting me know

**Comment by barrysteyn** [comment]
@BrandonZacharie Hi - this is resolved, tested and documented in 29dabe135c2423f3a173aebc6d7e2deaf8d145db

I realized that the initial fix I made for you was a bit buggy. So I did some more implementation, and added things to the tests. I also updated the documentation.

Please can you check the documentation and if you have anything to add, please fork and make a pull request.

Thanks so much for all the information and help and great debates :)
